### ISSUE TITLE
Non-version package specifiers don't work

### ISSUE THREAD

**Issue by rotu** [nodejs/corepack#313]
Corepack rejects valid [package specifiers](https://docs.npmjs.com/cli/v10/using-npm/package-spec?v=true) which are not versions.

For instance the following are valid specifiers which are more specific than semver but which nevertheless are rejected by `corepack`:
- `npm@github:npm/cli#c12ea07`
- `yarn@https://registry.npmjs.org/yarn/-/yarn-2.4.3.tgz`

output:
```
corepack use yarn@https://registry.npmjs.org/yarn/-/yarn-2.4.3.tgz
Usage Error: Tag not found (https://registry.npmjs.org/yarn/-/yarn-2.4.3.tgz)

$ corepack use <pattern>
```
```
corepack install --global yarn@https://registry.npmjs.org/yarn/-/yarn-2.4.3.tgz
Internal Error: ENOENT: no such file or directory, stat '/Users/dan/Source/corepack/yarn@https:/registry.npmjs.org/yarn/-/yarn-2.4.3.tgz'
Error: ENOENT: no such file or directory, stat '/Users/dan/Source/corepack/yarn@https:/registry.npmjs.org/yarn/-/yarn-2.4.3.tgz'
```

**Comment by rotu** [comment]
Originally reported in https://github.com/nodejs/corepack/pull/312#issuecomment-1752169602

**Comment by aduh95** [comment]
I think this is the documented behavior: https://github.com/nodejs/corepack/blob/fe3e5cd86c45db0d87c7fdea87d57d59b0bdcb78/README.md#L80-L82

Should we close as Won't fix?

**Comment by arcanis** [comment]
I don't think we should support git repositories (they are very complex to clone / install / pack), but it'd imo be reasonable to allow arbitrary urls (either `.tgz` or `.js`). Can be useful for internal mirrors.

**Comment by merceyz** [comment]
> but it'd imo be reasonable to allow arbitrary urls

As build metadata or in general?

I don't think we should allow using just a `name@url` pattern as it would break "stuff" checking the field to see which version is in use.

**Comment by arcanis** [comment]
They could still easily call `--version` to know this information though.

**Comment by BuildBackBuehler** [comment]
I suppose this is a differing issue, but the man pages of corepack (node v21+) state one can also use

>corepack install yarn@*

https://nodejs.org/api/corepack.html

To install the latest pnpm. That doesn't work for me. Neither does using * work inside my package.json. i.e.,

> "packageManager": "pnpm@*" 

I also tried latest or leaving out the version altogether and it wasn't accepted.

**Comment by rotu** [comment]
I think there are two issues:

- `corepack install yarn@*` is probably being expanded by your shell. Try `corepack install ‘yarn@*’` instead.
- the `package.json` `packageManager` field does not take a package specifier - it only takes a resolved version.

**Comment by aduh95** [comment]
I think this has been solved by https://github.com/nodejs/corepack/issues/359.
